### ISSUE TITLE
Perf improvements

### ISSUE THREAD

**Issue by NicoKiaru** [BIOP/ijp-kheops#12]
In the pyramidizer which requires two writers (the multi res one and the current level), the writing could happen in parallel, leading to a potential factor 2 improvement.

**Comment by NicoKiaru** [comment]
Assessed on master (0.20.2-SNAPSHOT). **The structure described here is still exactly as it was in 2022, but the expected gain is not a factor 2 — measured, it is about 1.25x, and only for images that actually get a pyramid.**

### The code is unchanged

In `OMETiffExporter.export()` every tile is still written twice, sequentially, on the single writing thread, for every level `r < nResolutionLevels - 1`:

```java
if (r < nResolutionLevels - 1) {
    currentLevelWriter.saveBytes(plane, computedBlocks.get(key), ...);  // temp file, uncompressed
}
writer.saveBytes(plane, computedBlocks.get(key), ...);                  // final file, LZW
```

### Measurement

I timed the two `saveBytes` calls separately (temporary probe, reverted) on the benchmark datasets of `BENCHMARKS.md`: 32 logical processors, NVMe, 1024 px tiles, LZW, uncompressed temporary files, reader pool 8, 31 worker threads, median of a warmed-up run.

| dataset | total | temp save | main save | waiting for tiles | best case, fully overlapped |
| --- | ---: | ---: | ---: | ---: | ---: |
| in-memory 8000x6000x3 uint16 | 6338 ms | 1559 ms (25%) | 4304 ms (68%) | 5% | **x1.32** |
| VSI `10x_01`, 7230x6151x3 uint16 | 6514 ms | 1406 ms (22%) | 4250 ms (65%) | 9% | **x1.27** |
| VSI RGB overview, 16172x6817 | 7473 ms | 1383 ms (19%) | 4854 ms (65%) | 11% | **x1.23** |
| CZI series 0 (single level) | 2066 ms | 0 | 2000 ms (97%) | 0% | x1.00 |
| LIF series 0 (single level) | 423 ms | 0 | 281 ms (66%) | 13% | x1.00 |

"Best case, fully overlapped" is `total / (total - sum over tiles of min(temp, main))`, i.e. the two writes cost `max(a, b)` instead of `a + b` per tile.

**Why it is not 2x:** the two writes are asymmetric. The main writer compresses (LZW) and carries the pyramid / SubIFD bookkeeping; the temporary writer writes raw bytes to a file that is deleted at the end. The measured ratio is about 3:1, so `(a+b)/max(a,b)` is ~1.3, not 2. An export with a single resolution level has no temporary writer at all and gains nothing.

Two caveats on the table. It is an **upper bound**: it assumes no queue-handoff cost and no contention between the two write streams, which holds on the NVMe measured here and much less on a spinning disk or on SMB, where the two streams would share the same bandwidth. And the low "waiting for tiles" column (5–11%) is the good news — the writing thread really is saturated, so the workers would not simply absorb the gain.

Implementation cost, for the record: a bounded queue plus a dedicated temp-writer thread, joined before `currentLevelWriter.close()`, with cancellation and exception propagation, and `computedBlocks.remove(key)` deferred until the temporary writer has consumed the array — which raises peak memory above `maxTilesInQueue`.

### The alternative, and its memory cost

The same 19–25% can be removed entirely rather than merely overlapped, by dropping the temporary file and downsampling from the tiles as they stream past. That also removes the temp-file reads and the per-thread `OMETiffReader` in `computeTile`, and it gets *better* on slow storage, where the parallel-write version gets worse.

The obvious objection is memory, so here are the numbers, since "safe on gigantic single planes" is a requirement.

A naive band — buffering two raw rows of level-`r` tiles — costs roughly `2 x width x tileY x bytesPerPixel`. At 100 000 px wide, uint16, `tileY = 1024`, that is **410 MB**. That objection is correct: the naive version does not scale to whole-slide widths.

But the band does not have to hold raw tiles. With `downsample = 2` and equal tile sizes across levels, the region a level-`r+1` tile needs is exactly a 2x2 group of level-`r` tiles, so each level-`r` tile can be averaged down to a quarter tile the moment it is written, and only the quarter tiles have to wait for their group to complete. One row of quarter tiles is `width x tileY x bytesPerPixel / 4` — **51 MB** at the same 100 000 px width, or ~68 MB summed over all levels, since each level costs a quarter of the one below.

For comparison, the default single-series path already holds `maxTilesInQueue = 64` tiles in `computedBlocks`, which is 134 MB at 1024 px uint16. So the band is the same order as a buffer the exporter already carries, and can be bounded the same way. Crucially it is **O(width), not O(width x height)**: a 100 000 x 100 000 plane costs the same band as a 100 000 x 2 000 one.

Two honest wrinkles. `adjustTileSize` picks a tile size per resolution level, so the 2x2 alignment above is usual but not guaranteed; where it does not hold, one extra buffered row is needed. And in the untiled path (`tileSize <= 0`, #31) a "tile" is already a whole plane, so the exporter materialises full planes in RAM today regardless.

### Recommendation

This is a real trade-off rather than a clear win either way:

- **Parallelising the two writers** (this issue) keeps the memory profile exactly as it is — everything stays bounded by the temporary file on disk — and buys ~1.25x on pyramidal exports, 1.00x on single-level ones.
- **Dropping the temporary file** buys more than that and scales better on slow storage, but introduces a buffer that grows with image width, ~50–70 MB at 100 000 px and uint16.

Neither is close to the biggest remaining win on files with many small planes, where the export is bound by TIFF IFD write syscalls rather than by pixels (see #37).

**Comment by NicoKiaru** [comment]
Implemented on branch [`parallel-temp-writer`](https://github.com/BIOP/ijp-kheops/tree/parallel-temp-writer) (e13acee), not merged.

`OMETiffExporter.AsyncTileWriter`: one daemon thread with a short queue takes the temporary write, drained at every resolution level boundary before the temporary file is closed and read back. ~120 lines, no change to what is held on disk or in RAM beyond at most 4 queued tile arrays.

| dataset | before | after | speedup |
| --- | ---: | ---: | ---: |
| in memory, 8000x6000 x 3 uint16 | 5370 ms | 4231 ms | **x1.27** |
| VSI `10x_01` | 5434 ms | 4405 ms | **x1.23** |
| VSI RGB overview | 6573 ms | 5589 ms | **x1.18** |
| CZI series 0 (single level) | 1826 ms | 1829 ms | x1.00 |
| LIF series 0 (single level) | 345 ms | 354 ms | x0.97 |

Pixels of every plane of every resolution level hash identically before and after on all five datasets; 34/34 JUnit tests pass. Details in `BENCHMARKS.md` on the branch.

**Correction to my previous comment.** I wrote that dropping the temporary file altogether "buys more than that". It does not, and it cannot be dropped: the final writer runs in `setWriteSequentially(true)` mode, so all of level 0 is written before any of level 1, and the level 1 pixels produced meanwhile have to be parked somewhere — for a multi plane stack that is a quarter of the whole dataset, not a band. What that redesign can do is store the *downsampled* level in the temporary file instead of a copy of the current one, which is 4x less temporary data, so ~1.2x on top of what is already gained — and it costs a full width band of the next level in RAM, `width x tileY x bytesPerPixel` at the peak, ~100 MB for a 100 000 px wide uint16 plane. A rewrite of the pyramid path for the same order of gain as the 120 lines above.

**Comment by NicoKiaru** [comment]
Done and merged to master (`7546a98..0405cb5`), five commits.

The original idea — running the two writers in parallel — is in, and it led to three more changes on the same code. Measured end to end, LZW, reader pool 8, always alternating the builds inside one session:

| dataset | before | after | speedup |
| --- | ---: | ---: | ---: |
| in memory, 8000x6000 x 3 uint16 | 5370 ms | ~1850 ms | **~x2.9** |
| VSI `10x_01` | 5434 ms | ~2400 ms | **~x2.3** |
| VSI RGB overview | 6573 ms | ~5230 ms | **~x1.26** |
| CZI series 0, single level | 1826 ms | ~330 ms | **~x5.5** |
| LIF series 0, single level | 345 ms | ~160 ms | **~x2.2** |

Pixels hash identically to the previous master on all five datasets, at every resolution level; 34/34 JUnit tests pass.

What went in:

1. **e13acee** — the temporary file is written on its own thread (`AsyncTileWriter`). This issue's proposal. **x1.18 to x1.27**, not the factor 2 assumed here: the two writes are not symmetric, the final one compresses and keeps the SubIFD bookkeeping while the temporary one wrote raw bytes, so the ceiling is `(a+b)/max(a,b)` ~ 1.3.
2. **2c73bd1** — the worker threads compress the tiles and hand them over through `loci.formats.ICompressedTileWriter`, so the writing thread only writes. **x1.22 to x5.77**. The single level files gain most because of a second effect: the bulk copy branch of `TiffSaver.writeImage` needs `tileH*tileW*channels*bpp == buf.length`, which a partial tile never satisfies, so every edge tile used to be copied and zero padded one `writeByte` at a time on the writing thread.
3. **7c030a0** — a tile is serialized once instead of twice: the temporary writer takes the same compressed buffer as the final one. **x1.15** on a writer bound export, and it dropped the writing thread's wait on the temporary writer from 54 % to 7 %.
4. **0405cb5** — `compressTemporaryFiles` removed. It traded a second compression for disk space; there is no second compression left, so it had nothing to choose. Gone from both builders and from the four commands that exposed it as `compress_temp_files`. Macros passing it still work, but code calling either builder method has to drop the call.

Two things worth knowing for whoever touches this next:

- **RGB above resolution level 0 keeps compressing on the writing thread.** `saveCompressedBytes` writes a single strip, and the exporter turns interleaving off above level 0, where TIFF wants one strip per sample. Written as one strip the IFD claims three tiles at offsets 0 - silently corrupt. That is why the RGB overview gains least.
- **The bottleneck has moved to the worker threads**: 63 % of an in-memory export is now the writing thread waiting for them, against 20 % before. The writer side is close to done.

`-Dkheops.precompress=false` gives the compression back to the writer, in case a future Bio-Formats changes what a pre-compressed tile has to contain — that part is not documented by the interface.

Full tables, the method, and the measurement caveats are in `BENCHMARKS.md`. Closing.
